### DISCUSSION TITLE
Versioning with a History Table in 2.0 and Composite Indexes

### DISCUSSION THREAD

**Discussion by tummus** [sqlalchemy/sqlalchemy#10913] [Usage Questions]
I'm working on migrating my team's version of `history_meta` to unblock our 2.0 migration but I noticed that composite indexes are now copied over to the history_table via `local_mapper.local_table.to_metadata`, creating a collision in composite index naming. 

I've adjusted our version of the code to append `_history_table` to the index naming to avoid errors that stem from this:
```
for index in history_table.indexes:
    if len(index.columns) > 1:
        index.name = index.name + "_history_table"
```

However, I wanted to double check if there are other options of addressing this.

Also not sure 100% if this an issue that's specific to our configuration but thought it'd be best to flag here just in case others run into this.

**Comment by CaselIT** [comment]
Hi

> I noticed that composite indexes are now copied over to the history_table via `local_mapper.local_table.to_metadata`, creating a collision in composite index naming.

as far as I know indexes have always been copied by `to_metadata`. This is the code of 1.4
https://github.com/sqlalchemy/sqlalchemy/blob/0653c1ca11f4f5ac8ee17144d78afe71740a88e8/lib/sqlalchemy/sql/schema.py#L1141-L1155
On version 2 that code is basically the same

**Reply by tummus** [reply]
Makes sense, though might be good to flag add as a caveat in the 2.0 `history_meta` since it wasn't immediately obvious that I would need to handle this behavior since `to_metadata` wasn't in the 1.4 version of `history_meta`

**Reply by CaselIT** [reply]
Oh that's what changed. cc @zzzeek

**Comment by zzzeek** [comment]
I didnt understand this so I had to create a test to understand what the problem was.  this is a bug, will be fixed in 2.0.26 but of course this is an example so you'll be able to do whatever I come up with ahead of time.

**Comment by zzzeek** [comment]
here's your fix

```diff
diff --git a/examples/versioned_history/history_meta.py b/examples/versioned_history/history_meta.py
index 806267cb4..3f26832b9 100644
--- a/examples/versioned_history/history_meta.py
+++ b/examples/versioned_history/history_meta.py
@@ -56,6 +56,9 @@ def _history_mapper(local_mapper):
             local_mapper.local_table.metadata,
             name=local_mapper.local_table.name + "_history",
         )
+        for idx in history_table.indexes:
+            if idx.name is not None:
+                idx.name += "_history"
 
         for orig_c, history_c in zip(
             local_mapper.local_table.c, history_table.c

```

**Reply by tummus** [reply]
Thanks! Also not sure if there are other Index issues but I would assume unique indices would also have a problem with getting copied over (and would need to be set to non-unique)

**Reply by zzzeek** [reply]
I dont really understand how we aren't getting any reports of this.  every table has a PrimaryKeyConstraint.   All of these constraints have a name of None by default, but if a naming convention is in use, then they get a name, and im not sure if that updates when it copies.

this is all very strange because this is a popular recipe, naming conventions are popular, etc.  this should have been reported very early.

**Reply by zzzeek** [reply]
unique constraints do not have a problem.   which is...because we throw them out:

```
        for const in list(history_table.constraints):
            if not isinstance(
                const, (PrimaryKeyConstraint, ForeignKeyConstraint)
            ):
                history_table.constraints.discard(const)
```

so lets try FKConstraints...or naming convention with PKs..

**Reply by zzzeek** [reply]
for FK constraints and likely PK constraints, the names do in fact overlap.  however, SQlite, PostgreSQL, Oracle, SQL Server, MySQL, MariaDB, all accept this - these names are local to the parent table.
